FWCore/Concurrency/interface/Async.h

```diff
@@ -0,0 +1,28 @@
+#ifndef FWCore_Concurrency_Async_h
+#define FWCore_Concurrency_Async_h
+
+#include "FWCore/Concurrency/interface/WaitingTaskWithArenaHolder.h"
+#include "FWCore/Concurrency/interface/WaitingThreadPool.h"
+
+namespace edm {
+  // All member functions are thread safe
+  class Async {
+  public:
+    Async() = default;
+    virtual ~Async();
```

Comment:
Should this be marked `noexcept`?

Reply:
It would probably be good. Looking what gets done by the destructor, presently the destructor of `edm::ReusableObjectHolder` is not `noexcept`. Could or should it be?
 https://github.com/cms-sw/cmssw/blob/774f85421329c42cd1d30ec43d470f527141fb92/FWCore/Utilities/interface/ReusableObjectHolder.h#L91-L97
The OneAPI spec doesn't say anything about exceptions in `oneapi::tbb::concurrent_queue<...>::try_pop()` https://spec.oneapi.io/versions/latest/elements/oneTBB/source/containers/concurrent_queue_cls/safe_member_functions.html#popping-elements
so I guess the question is what we want the behavior to be if the destructor of `T` throws.

Well, if the destructor of `T` would throw, the `while` loop would terminate early and leave the remaining `T` to not be destructed, which is probably bad enough to just call `std::terminate` (i.e. suggesting to mark the `ReusableObjectHolder` destructor as `noexcept`.